apps/3d_rec_framework/src/tools/global_classification.cpp

```diff
@@ -5,235 +5,257 @@
-      std::dynamic_pointer_cast<pcl::rec_3d_framework::CVFHEstimation<pcl::PointXYZ, pcl::VFHSignature308>> (vfh_estimator)
-    );
-
-    pcl::rec_3d_framework::GlobalNNPipeline<Metrics::HistIntersectionUnionDistance, pcl::PointXYZ, pcl::VFHSignature308> global;
-    global.setDataSource (cast_source);
-    global.setTrainingDir (training_dir);
-    global.setDescriptorName (desc_name);
-    global.setFeatureEstimator (cast_estimator);
-    global.setNN (NN);
-    global.initialize (false);
-
-    segmentAndClassify<Metrics::HistIntersectionUnionDistance, pcl::PointXYZ, pcl::VFHSignature308> (global);
+  if (desc_name == "cvfh") {
+    std::shared_ptr<
+        pcl::rec_3d_framework::CVFHEstimation<pcl::PointXYZ, pcl::VFHSignature308>>
+        vfh_estimator(new pcl::rec_3d_framework::CVFHEstimation<pcl::PointXYZ,
+                                                                pcl::VFHSignature308>);
+    vfh_estimator->setNormalEstimator(normal_estimator);
+
+    std::shared_ptr<
+        pcl::rec_3d_framework::GlobalEstimator<pcl::PointXYZ, pcl::VFHSignature308>>
+        cast_estimator(std::dynamic_pointer_cast<
+                       pcl::rec_3d_framework::CVFHEstimation<pcl::PointXYZ,
+                                                             pcl::VFHSignature308>>(
+            vfh_estimator));
+
+    pcl::rec_3d_framework::GlobalNNPipeline<Metrics::HistIntersectionUnionDistance,
+                                            pcl::PointXYZ,
+                                            pcl::VFHSignature308>
+        global;
+    global.setDataSource(cast_source);
+    global.setTrainingDir(training_dir);
+    global.setDescriptorName(desc_name);
+    global.setFeatureEstimator(cast_estimator);
+    global.setNN(NN);
+    global.initialize(false);
+
+    segmentAndClassify<Metrics::HistIntersectionUnionDistance,
+                       pcl::PointXYZ,
+                       pcl::VFHSignature308>(global);
   }
 
-  if (desc_name == "esf")
-  {
-    std::shared_ptr<pcl::rec_3d_framework::ESFEstimation<pcl::PointXYZ, pcl::ESFSignature640>> estimator (
-      new pcl::rec_3d_framework::ESFEstimation<pcl::PointXYZ, pcl::ESFSignature640>
-    );
-
-    std::shared_ptr<pcl::rec_3d_framework::GlobalEstimator<pcl::PointXYZ, pcl::ESFSignature640>> cast_estimator (
-      std::dynamic_pointer_cast<pcl::rec_3d_framework::ESFEstimation<pcl::PointXYZ, pcl::ESFSignature640>> (estimator)
-    );
-
-    pcl::rec_3d_framework::GlobalNNPipeline<flann::L1, pcl::PointXYZ, pcl::ESFSignature640> global;
-    global.setDataSource (cast_source);
-    global.setTrainingDir (training_dir);
-    global.setDescriptorName (desc_name);
-    global.setFeatureEstimator (cast_estimator);
-    global.setNN (NN);
-    global.initialize (false);
-
-    segmentAndClassify<flann::L1, pcl::PointXYZ, pcl::ESFSignature640> (global);
+  if (desc_name == "esf") {
+    std::shared_ptr<
```

Comment:
It looks ugly - yes. But I think it is better to look at this in a separate PR, when the whole code will be maybe nested into a
```
namespace pcl {
namespace rec_3d_framework {
...
}
}
```
...as I don't really want touch code in such a PR.

Reply:
Added a slightly improvement to it by using `std::make_shared` and `auto`. We could even avoid the `dynamic_pointer_cast` by using:
```
    std::shared_ptr<
        pcl::rec_3d_framework::GlobalEstimator<pcl::PointXYZ, pcl::ESFSignature640>>
        estimator =
            std::make_shared<pcl::rec_3d_framework::
                                 ESFEstimation<pcl::PointXYZ, pcl::ESFSignature640>>();
```
but this looks worse as current code after applying clang-format, so this can maybe be done after changing the namespace issue mentioned above.